### PR TITLE
Enable to set unused joints for IK

### LINKED ISSUE #36: Support unused joints
Sometimes we want to fix some joints when solving IK.
For example, don't use wheel base for small manipulation.

### PR DESCRIPTION
**Pull request by ssr-yuki**
closes #36 

If unused joints are set, in IK calculation, 

1. the gradient on each step is projected into a feasible space
2. a linear equation is solved numerically
3. zero-gradient corresponding to unused joint is recovered to update joint angles

You can check the behavior with the following command:
```
cargo run --release --example interactive_ik -- -i shoulder_pitch
```
In this case, `shoulder_pitch` joint is fixed.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,7 @@ serde-serialize = ["nalgebra/serde-serialize", "serde"]
 log = "0.4"
 nalgebra = "0.26"
 simba = "0.4"
+structopt = "0.3"
```
Comment (OTL) [inline Cargo.toml:22]: Move to `dev-dependencies`.

File: src/ik.rs [hunk 2]
```diff
@@ -215,39 +215,71 @@ where
...
         for (i, use_i) in operational_space.iter().enumerate() {
             if !use_i {
-                jacobi = jacobi.remove_row(i - removed_count);
-                removed_count += 1;
+                jacobi = jacobi.remove_row(i - num_removed_rows);
+                num_removed_rows += 1;
+            }
+        }
+
+        let mut ignored_joint_indices = Vec::<usize>::new();
```
Comment (OTL) [inline src/ik.rs:233]: I think you can remove `::<usize>`

```uggestion
        let mut ignored_joint_indices = Vec::new();
```

File: src/ik.rs [hunk 3]
```diff
@@ -215,39 +215,71 @@ where
...
+        for joint_name in constraints.ignored_joint_names.iter() {
+            // Try to get joint index
+            match arm.iter_joints().position(|x| x.name == *joint_name) {
```
Comment (OTL) [inline src/ik.rs:236]: Finding string from Vec<String> is slow. I want to this logic to out of the loop.
If you can change the interface of `solve_one_loop_with_constraints`, is it possible to move out this logic?

**Review by OTL — COMMENTED** [review]
And I'm happy if you add some easy (not so serious) tests to `tests/test_ik.rs`

File: src/ik.rs [hunk 4]
```diff
@@ -212,42 +212,58 @@ where
...
+
+                    let mut subtask = na::DVector::from_vec(f(&orig_positions));
+                    for (i, joint_index) in ignored_joint_indices.iter().enumerate() {
+                        subtask = subtask.remove_row(*joint_index - i);
+                    }
+                    let mut d_q = jacobi_inv.clone() * err
+                        + (na::DMatrix::identity(available_dof, available_dof)
+                            - jacobi_inv * jacobi)
+                            * subtask;
+                    for joint_index in ignored_joint_indices.iter() {
```
Comment (OTL) [inline src/ik.rs:252]: We don't need `iter()`

```suggestion
                    for joint_index in ignored_joint_indices {
```

File: src/ik.rs [hunk 5]
```diff
@@ -212,42 +212,58 @@ where
...
-                    jacobi
+                None => {
+                    let mut d_q = jacobi
                         .svd(true, true)
                         .solve(&err, na::convert(EPS))
-                        .unwrap() // TODO
-                        .as_slice(),
-                ),
+                        .unwrap();
+                    for joint_index in ignored_joint_indices.iter() {
```
Comment (OTL) [inline src/ik.rs:262]: ditto

File: src/ik.rs [hunk 6]
```diff
@@ -275,18 +291,34 @@ where
...
+        if available_dof < required_dof {
             return Err(Error::PreconditionError {
-                dof: orig_positions.len(),
-                necessary_dof: use_dof,
+                dof: available_dof,
+                necessary_dof: required_dof,
             });
         }
+        let mut ignored_joint_indices = Vec::new();
+        for joint_name in constraints.ignored_joint_names.iter() {
```
Comment (OTL) [inline src/ik.rs:304]: We can use this if you want.

```suggestion
        for joint_name in &constraints.ignored_joint_names {
```

**Review by OTL — APPROVED** [review]

**Comment by ssr-yuki** [comment]
Thank you for reviewing!

**Review by OTL — APPROVED** [review]
👍🏼👍🏼👍🏼

**Pull request merged**